httpbakery/form/form_test.go

```diff
@@ -0,0 +1,219 @@
+	expectError: `cannot get discharge from ".*": cannot start interactive session: cannot get login methods: GET .*: httprequest: test error`,
+}, {
+	about:        "login methods not supported",
+	ignoreAccept: true,
+	expectError:  `cannot get discharge from ".*": cannot start interactive session: cannot get login methods: GET .*: unexpected content type text/plain; want application/json; content: OK`,
+}, {
+	about:           "form login method not supported",
+	formUnsupported: true,
+	expectError:     `cannot get discharge from ".*": cannot start interactive session: form login not supported`,
+}, {
+	about:       "error getting schema",
+	getError:    true,
+	expectError: `cannot get discharge from ".*": cannot start interactive session: cannot get schema: GET .*: httprequest: test error`,
+}, {
+	about:       "error submitting form",
+	postError:   true,
+	expectError: `cannot get discharge from ".*": cannot start interactive session: cannot submit form: POST .*: httprequest: test error`,
+}, {
+	about:       "no schema",
+	emptySchema: true,
+	expectError: `cannot get discharge from ".*": cannot start interactive session: invalid schema`,
+}, {
+	about: "handler error",
+	handler: func(environschema.Fields) (map[string]interface{}, error) {
+		return nil, testError
+	},
+	expectError: `cannot get discharge from ".*": cannot start interactive session: cannot handle form: test error`,
+}}
+
+func (s *formSuite) TestFormLogin(c *gc.C) {
+	d := &formDischarger{}
+	d.discharger = bakerytest.NewLoginDischarger(nil, http.HandlerFunc(d.login))
+	defer d.discharger.Close()
+	d.discharger.Mux.Handle("/form", http.HandlerFunc(d.form))
+	svc, err := bakery.NewService(bakery.NewServiceParams{
+		Locator: d.discharger,
+	})
+	c.Assert(err, gc.IsNil)
+	for i, test := range formLoginTests {
+		c.Logf("%d. %s", i, test.about)
+		d.ignoreAccept = test.ignoreAccept
+		d.loginError = test.loginError
+		d.formUnsupported = test.formUnsupported
+		d.getError = test.getError
+		d.postError = test.postError
+		d.emptySchema = test.emptySchema
+
+		m, err := svc.NewMacaroon("", nil, []checkers.Caveat{{
+			Location:  d.discharger.Location(),
+			Condition: "test condition",
+		}})
+		c.Assert(err, gc.Equals, nil)
+		client := httpbakery.NewClient()
+		h := defaultHandler
+		if test.handler != nil {
+			h = test.handler
+		}
+		form.SetUpAuth(client, h)
+
+		ms, err := client.DischargeAll(m)
+		if test.expectError != "" {
+			c.Assert(err, gc.ErrorMatches, test.expectError)
+			continue
+		}
+		c.Assert(err, gc.IsNil)
+		c.Assert(len(ms), gc.Equals, 2)
+	}
+}
+
+func (s *formSuite) TestFormLoginNewRequestError(c *gc.C) {
+	client := httpbakery.NewClient()
+	form.SetUpAuth(client, defaultHandler)
+	u := url.URL{
+		Scheme: ":",
+	}
+	err := client.VisitWebPage(&u)
+	c.Assert(err, gc.ErrorMatches, "cannot create request: parse :://: missing protocol scheme")
+}
+
+type formDischarger struct {
+	discharger *bakerytest.LoginDischarger
+
+	ignoreAccept    bool
```

Comment:
How about putting all these fields in a struct type so that we
can copy the whole lot easily in the test. Makes the tests
slightly more verbose, but probably worth it.